examples/hpc/dla-driver-ffi/src/lib.rs

```diff
@@ -333,7 +346,139 @@ pub unsafe extern "C" fn dla_conv2d_bias_relu(
+    input_scale: *const f32,
+    input_zero: *const i32,
+    input_channels: usize,
+    input_height: usize,
+    input_width: usize,
+    input_order: *const c_char,
+    kernel_amount: usize,
+    kernel_channels: usize,
+    kernel_height: usize,
+    kernel_width: usize,
+    kernel_order: *const c_char,
+    bias_length: usize,
+    pad_top: u32,
+    pad_right: u32,
+    pad_left: u32,
+    pad_bottom: u32,
+    pad_value: i32,
+    stride_x: u32,
+    stride_y: u32,
+    mac_clip: u32,
+    pp_clip: u32,
+) {
+    let input_scale: Vec<f32> =
+        unsafe { slice::from_raw_parts(input_scale as *const f32, 1).to_vec() };
+
+    let input_zero: Vec<i32> =
+        unsafe { slice::from_raw_parts(input_zero as *const i32, 1).to_vec() };
+
+    let output_scale: Vec<f32> =
+        unsafe { slice::from_raw_parts(output_scale as *const f32, kernel_amount).to_vec() };
+
+    let output_zero: Vec<i32> =
+        unsafe { slice::from_raw_parts(output_zero as *const i32, 1).to_vec() };
+
+    let (input_tensor, kernels_tensor) = unsafe {
+        ffi_data_import(
+            input_data,
+            input_channels,
+            input_height,
+            input_width,
+            input_order,
+            (-1 * input_zero[0]) as i16,
+            kernel_data,
+            kernel_amount,
+            kernel_channels,
+            kernel_height,
+            kernel_width,
+            kernel_order,
+        )
     };
+
+    let bias: Vec<i32> = unsafe { slice::from_raw_parts(bias as *const i32, bias_length).to_vec() };
+    let bias_i16: Vec<i16> = bias.into_iter().map(|x| clip_i32_to_i16(x)).collect();
+
+    let mut result = conv2d_bias_relu(
+        input_tensor,
+        kernels_tensor,
+        bias_i16,
+        Some(Padding {
+            top: pad_top,
+            right: pad_right,
+            left: pad_left,
+            bottom: pad_bottom,
+            padding_value: pad_value,
+        }),
+        Some(Stride {
+            x: stride_x,
+            y: stride_y,
+        }),
+        Some(mac_clip),
+        Some(pp_clip),
+        None,
+    );
+
+    // TVM requantization and clip
+    rescale(
+        &mut result,
+        u32::pow(2, pp_clip) as f32, //NOTE:(20240924 vaino-waltteri.granat@tuni.fi) Mitigate pp downscale
+        input_zero[0],
+        output_zero[0],
+        input_scale[0],
+        output_scale,
+    );
+
+    let input_order_string = unsafe { CStr::from_ptr(input_order).to_str().unwrap_unchecked() };
+
+    unsafe {
+        core::ptr::copy_nonoverlapping(
+            result
+                .to_buffer_with_order(Order3::try_from(input_order_string).unwrap_unchecked())
+                .as_mut_ptr(),
+            output,
+            result.get_size(),
+        )
+    };
+}
+fn clip_i32_to_i16(value: i32) -> i16 {
+    if value > i16::MAX as i32 {
+        return i16::MAX;
+    } else if value < i16::MIN as i32 {
+        return i16::MIN;
+    } else {
+        return value as i16;
+    }
```

Comment:
Consider the standard library function `value.clamp(i16::MIN as i32, i16::MAX as i32) as i16` which implements the same behavior. It does include "an extra" assert for checking `MAX > MIN`, but I'm quite confident the compiler gets rid of that since MIN & MAX are known at compile-time.

Reply:
Function removed and replaced by direct calls to clamps